### DISCUSSION TITLE
Using git submodules with GitHub Pages and Jekyll

### DISCUSSION THREAD

**Discussion by benfrancis** [community/community#23634] [Other Feature Feedback, Questions, & Ideas]
<p>Hi,</p>
<p>I have a website hosted on GitHub Pages and am trying to add a sub-folder for that site as a git submodule, which uses Jekyll to create a themed sub-site.</p>
<p>GitHub’s documentation <a href="https://docs.github.com/en/free-pro-team@latest/github/working-with-github-pages/using-submodules-with-github-pages" rel="noopener nofollow ugc">says</a>:</p>
<blockquote>
<p>If the repository for your GitHub Pages site contains submodules, their contents will automatically be pulled in when your site is built.</p>
</blockquote>
<p>This does seem to work and the Markdown content in my submodule gets rendered as HTML, but the _config.yml configuration file inside the submodule seems to be ignored and its jekyll theme is not applied.</p>
<p>Is there a way to get the jekyll configuration file inside the submodule to be used? Perhaps with a top level jekyll configuration file? If that is the solution then what happens if I have multiple sub-modules with different jekyll configurations?</p>
<p>Thanks</p>

**Comment by otter-computer** [comment]
<p>Hi <a class="mention" href="/benfrancis">@benfrancis</a>, welcome to the GitHub Support Community!</p>
<p>Jekyll only supports a single root-level <code>_config.yml</code> file per site. This means that Pages won’t take into account the config file inside your submodule.</p>
<p>Can you give some more information on exactly why you want to use multiple? Are you importing a site from another repository in order to have it hosted on GitHub Pages too? If so, it may be simpler to just enable Pages on the repository you’re importing as a submodule instead.</p>

**Comment by benfrancis** [comment]
<p>Hi <a class="mention" href="/thomasshaped">@thomasshaped</a>, thanks for your response.</p>
<p>The top level site is a static HTML website which does not use Jekyll, but the /docs subdirectory (which is a separate repo included as a submodule) does. We previously used Netlify to deploy the site and its subdirectories but have moved back to GitHub Pages using submodules and would ideally like other submodules to use Jekyll as well (e.g. /blog).</p>
<p>Top level site repo: <a href="https://github.com/WebThingsIO/webthings.io" rel="noopener nofollow ugc">https://github.com/WebThingsIO/webthings.io</a><br>
Docs submodule repo: <a href="https://github.com/WebThingsIO/docs" rel="noopener nofollow ugc">https://github.com/WebThingsIO/docs</a></p>
<p>I tried adding a _config.yml file to the top level repository which pointed to /docs as a source and target for jekyll, but it didn’t work. Is there a way to make this work? Presumably from what you’ve said if it does work it will only work for a single subdirectory/submodule?</p>
<p>We are considering switching /docs from Jekyll to MkDocs so if we do that then supporting a single submodule for /blog might be enough.</p>
<p>Thanks</p>

**Comment by otter-computer** [comment, marked as answer]
<p>Thanks for explaining that, makes more sense to me what you’re doing now!</p>
<p>Pages will always only take into account a root-level <code>_config.yml</code> file. This means that any config files inside your submodule repositories will be ignored when they are imported, meaning that any specific themes you’ve chosen will be ignored.</p>
<p>If you’re switching to MKDocs then that should solve your problem just fine for this issue, but you may have to make certain considerations if you’re planning to use Jekyll again in future. For example, any Jekyll configuration will need to be made in the <code>WebThingsIO/webthings.io</code> repository, regardless if you’re using Jekyll for that direct site or not, as the root of that repository is the only place where Pages will look for a config file.</p>
<p>Another alternative would be to host separate Pages sites out of each of your repositories.</p>
<p>I notice that you’re using the <code>webthings.io</code> custom domain for this site. To keep your existing URLs the same you’ll need to rename your repository from <code>WebThingsIO/webthings.io</code> to <code>WebThingsIO/webthingsio.github.io</code>. Doing this will allow this custom domain to be inherited by all other Pages sites listed under your account instead of being hosted from a <code>github.io</code> address.</p>
<p>Once you’ve made that change you should then be able to go to any other repositories hosted under your GitHub account an enable Pages in their repository settings. Any repositories you do this with will be available at <a href="https://webthings.io/repository-name/">https://webthings.io/repository-name/</a>. If you do this with your <code>WebThingsIO/docs</code> repository your site will then be hosted on Pages, with the correct Jekyll theme, at <a href="https://webthings.io/docs/">https://webthings.io/docs/</a>.</p>
<p>If you’re unhappy changing the name of your <code>WebThingsIO/webthings.io</code> repository then you can still do this, though any existing domains will no longer work and Pages sites outside of this repository will be hosted from a <code>github.io</code> address (unless you add a different custom domain/subdomain to them in the repository settings).</p>
<p>You can choose to mix and match this approach—where some repositories are included as submodules and others are hosted as separate sites—but any Jekyll site will need to be hosted as a separate repository or will need to rely on the config file from the root repository.</p>
<p>[Edited to fix an incorrect repository name]</p>

**Comment by benfrancis** [comment]
<p>Fantastic, thanks! I totally forgot about the magic .github.io repo name and didn’t realise it would have that effect.</p>
<p>Marking this as solution but one slight edit, the repo name should be <code>WebThingsIO/webthingsio.github.io</code> not <code>WebThingsIO/webthings.github.io</code></p>

**Comment by otter-computer** [comment]
<p>Oops, I missed that, sorry! You’re right, that is the correct repo name, it should use the org/account name :sweat_smile:</p>
